src/agents/der_aggregator.jl

```diff
@@ -0,0 +1,636 @@
+abstract type AbstractDERA <: Agent end
+
+struct DERAOptions <: AgentOptions
+    solvers::HEMSolver
+    # solvers::Union{HEMSolver, Dict{String, <:HEMSolver}}
+end
+
+"""
+Construct DERAOptions with an MOI.OptimizerWithAttributes instance.
+"""
+function DERAOptions(attributes::MOI.OptimizerWithAttributes)
+    return DERAOptions(AnySolver(attributes))
+end
+
+mutable struct DERA <: AbstractDERA
+    id::String
+    # incentive function for DER aggregation (piece-wise linear function: incentive (x) vs participation (y))
+    dera_stor_incentive_function::DataFrame
+    # parameters applied to aggregated capacities of DER to represent potential friction of aggregation
+    aggregation_friction::ParamArray
+    # incentive level by year
+    incentive_level::ParamArray
+    # aggregation level by year
+    aggregation_level::ParamArray
+    # percentage of cost savings as revenues for DERA under VIU
+    rev_perc_cost_saving_viu::ParamArray
+    dera_stor_my::ParamArray
+    dera_pv_my::ParamArray
+end
+
+function DERA(input_filename::AbstractString, model_data::HEMData; id = DEFAULT_ID)
+
+    # need to have the incentive function for each customer type
+    dera_stor_incentive_function = CSV.read(joinpath(input_filename, "dera_stor_incentive_function.csv"), DataFrame)
+
+    return DERA(
+        id,
+        dera_stor_incentive_function,
+        initialize_param("aggregation_friction", model_data.index_y, model_data.index_z, model_data.index_h, value = 0.0),
+        initialize_param("incentive_level", model_data.index_y, model_data.index_z, value = 0.0),
+        initialize_param("aggregation_level", model_data.index_y, model_data.index_z, value = 0.0),
+        initialize_param("rev_perc_cost_saving_viu", model_data.index_y, model_data.index_z, value = 0.1),
+        initialize_param("dera_stor_my", model_data.index_y, model_data.index_z, model_data.index_h, value = 0.0),
+        initialize_param("dera_pv_my", model_data.index_y, model_data.index_z, model_data.index_h, value = 0.0),
+    )
+end
+
+get_id(x::DERA) = x.id
+
+function solve_agent_problem!(
+    der_aggregator::DERA,
+    dera_opts::AgentOptions,
+    model_data::HEMData,
+    hem_opts::HEMOptions{WholesaleMarket},
+    agent_store::AgentStore,
+    w_iter,
+    jump_model,
+    export_file_path
+)
+
+    reg_year = model_data.year(first(model_data.index_y))
+    reg_year_index = Symbol(Int(reg_year))
+    delta_t = parse(Int64, chop(string(model_data.index_t.elements[2]), head = 1, tail = 0)) - parse(Int64, chop(string(model_data.index_t.elements[1]), head = 1, tail = 0))
+
+    ipp = get_agent(IPPGroup, agent_store)
+    customers = get_agent(CustomerGroup, agent_store)
+
+    # the x-axis of incentive function is assumed to be distributed storage capacity (even though the technology is PV+storage)
```

Comment:
Capacity in MW, or percent of customers participating?

Reply:
This is another bad comment from me. What I meant was the x-axis of the incentive function is $ per MW of distributed storage capacity (per year). The percentage of customers participating is y-axis of the incentive function.

Reply:
Nongchao to revise the comment